src/python/pants/backend/python/framework/stevedore/python_target_dependencies.py

```diff
@@ -0,0 +1,222 @@
+from pants.engine.unions import UnionRule
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+from pants.util.ordered_set import OrderedSet
+from pants.util.strutil import softwrap
+
+# -----------------------------------------------------------------------------------------------
+# Utility rules to analyze all StevedoreNamespace entry_points
+# -----------------------------------------------------------------------------------------------
+
+
+@rule(
+    desc=f"Find all `{PythonDistribution.alias}` targets with `{StevedoreNamespace.alias}` entry_points",
+    level=LogLevel.DEBUG,
+)
+def find_all_python_distributions_with_any_stevedore_entry_points(
+    targets: AllTargets,
+) -> AllStevedoreExtensionTargets:
+    return AllStevedoreExtensionTargets(
+        tgt
+        for tgt in targets
+        if tgt.has_field(PythonDistributionEntryPointsField)
+        and any(
+            # namespace aka category aka group
+            isinstance(namespace, StevedoreNamespace)
+            for namespace in (tgt[PythonDistributionEntryPointsField].value or {}).keys()
+        )
+    )
+
+
+@dataclass(frozen=True)
+class StevedoreExtensions:
+    """A mapping of stevedore namespaces to a list of targets that provide them.
+
+    Effectively, the targets are StevedoreExtension targets.
+    """
+
+    mapping: FrozenDict[str, tuple[Target, ...]]
+
+
+@rule(
+    desc=f"Create map of `{StevedoreNamespace.alias}` to `{PythonDistribution.alias}` targets",
+    level=LogLevel.DEBUG,
+)
+async def map_stevedore_extensions(
+    stevedore_extensions: AllStevedoreExtensionTargets,
+) -> StevedoreExtensions:
+    mapping: Mapping[str, list[Target]] = defaultdict(list)
+    for tgt in stevedore_extensions:
+        # namespace aka category aka group
+        for namespace, entry_points in (
+            tgt[PythonDistributionEntryPointsField].value or {}
+        ).items():
+            if isinstance(namespace, StevedoreNamespace):
+                mapping[str(namespace)].append(tgt)
+    return StevedoreExtensions(FrozenDict((k, tuple(v)) for k, v in sorted(mapping.items())))
+
+
+@rule(
+    desc=f"Find `{PythonDistribution.alias}` targets with entry_points in selected `{StevedoreNamespace.alias}`s",
+    level=LogLevel.DEBUG,
+)
+def find_python_distributions_with_entry_points_in_stevedore_namespaces(
+    request: StevedoreNamespacesProviderTargetsRequest,
+    stevedore_extensions: StevedoreExtensions,
+) -> StevedoreExtensionTargets:
+    namespaces: StevedoreNamespacesField = request.stevedore_namespaces
+    if namespaces.value is None:
+        return StevedoreExtensionTargets(())
+
+    return StevedoreExtensionTargets(
+        {
+            tgt
+            for namespace in namespaces.value
+            for tgt in stevedore_extensions.mapping.get(namespace, ())
+            # We only allow defining the stevedore namespaces in the entry_points field.
+            if tgt[PythonDistributionEntryPointsField].value
```

Comment:
I don't see how this `if` guard would ever be false? The `tgt` wouldn't be in the mapping if it's entry points field didn't have a value.. (I don't get the relevance of the comment either :) )

Reply:
Ah. True.

I'll try removing that to see if MyPy complains.

The comment probably goes somewhere else. On a `python_distribution`, entry_points can be defined in both the `entry_points={...}` and the `provides=python_artifact(entry_points={...})` fields. This comment was meant to clarify that I chose to only support the `entry_points=` field, not the `provides=` field.

Reply:
I moved the comment up higher and expanded it.

Reply:
Ah, gotcha. Yea that wasn't clear to me :)